lib/internal/modules/run_main.js

```diff
@@ -92,10 +93,24 @@ function shouldUseESMLoader(mainPath) {
  */
 function runMainESM(mainPath) {
   const { loadESM } = require('internal/process/esm_loader');
-  const { pathToFileURL } = require('internal/url');
+  const { pathToFileURL, URL } = require('internal/url');
+  const _importMapPath = getOptionValue('--experimental-import-map');
   const main = pathToFileURL(mainPath).href;
 
   handleMainPromise(loadESM((esmLoader) => {
+    // Load import map and throw validation errors
+    if (_importMapPath) {
+      const { ImportMap } = require('internal/modules/esm/import_map');
+      const { getCWDURL } = require('internal/util');
+
+      const importMapPath = esmLoader.resolve(_importMapPath, getCWDURL(), { __proto__: null, type: 'json' });
+      return esmLoader.import(importMapPath.url, getCWDURL(), { __proto__: null, type: 'json' })
+        .then((importedMapFile) => {
+          esmLoader.importMap = new ImportMap(importedMapFile.default, new URL(importMapPath.url));
+          return esmLoader.import(main, undefined, { __proto__: null });
+        });
```

Comment:
Calling `.resolve` then `.import` will effectively call `resolve` twice. We can avoid that:
```suggestion
      const parentURL = getCWDURL();
      const importAttributes = { __proto__: null, type: 'json' };
      const importMapPath = await esmLoader.resolve(_importMapPath, parentURL, importAttributes);
      const job = await esmLoader.getJobFromResolveResult(importMapPath, parentURL, importAttributes);
      const { module: importMapModule } = await job.run();
      esmLoader.importMap = new ImportMap(importMapModule.getNamespace().default, new URL(importMapPath.url));
```

Reply:
Cool, this is what I was asking about in the other PR on my fork. I was pretty sure there must be a better way here. Will update.